### PR TITLE
Re train cron setup

### PR DESCRIPTION
**Pull request by krishrustagi**

### CODE REVIEW

File: emotunes-admin/src/main/java/com/emotunes/emotunes/service/impl/AdminServiceImpl.java [hunk 1]
```diff
@@ -210,11 +212,11 @@ private void availAllSongsToUser(String userId, String modelWeightsUrl) {
 
     private void predictEmotionAndPersistMapping(
             String userId, String songId, String songUrl, String modelWeightsUrl) {
-        LinkedMultiValueMap<String, String> multiValueMap = new LinkedMultiValueMap<>();
-        multiValueMap.add("song_url", songUrl);
-        multiValueMap.add("model_weights_url", modelWeightsUrl);
+        Map<String, String> map = new HashMap<>();
+        map.put("song_url", songUrl);
```
Comment (SarthakSavasil) [inline emotunes-admin/src/main/java/com/emotunes/emotunes/service/impl/AdminServiceImpl.java]: Kaam kr gya?

File: emotunes-feign/src/main/java/com/emotunes/emotunes/client/MachineLearningClient.java [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
+import java.util.Map;
+
 @FeignClient(name = "machineLearningClient", url = "https://emtounes.onrender.com/")
 public interface MachineLearningClient {
 
     @PostMapping(value = "/v1/predict")
-    String predictEmotion(@RequestBody LinkedMultiValueMap<String, String> formData);
+    String predictEmotion(@RequestBody Map<String, String> formData);
+
+    @PostMapping(value = "v1/re_train")
```
Comment (SarthakSavasil) [inline emotunes-feign/src/main/java/com/emotunes/emotunes/client/MachineLearningClient.java]: underscore use mt kr hiphen kr

File: emotunes-commons/src/main/java/com/emotunes/emotunes/repository/UserSongEmotionPreferenceRepository.java [hunk 3]
```diff
@@ -2,8 +2,19 @@
...
 @Repository
 public interface UserSongEmotionPreferenceRepository extends JpaRepository<StoredUserSongEmotionPreference, String> {
+
+    @Query(value =
+            "select user_id, song_id from user_song_emotion_preference"
+                    + " group by user_id"
+                    + " having count(user_id) > ?1"
+                    + " order by user_id",
+            nativeQuery = true)
+    List<List<String>> findUserIdSongIdWithUserCountLimit(long limit);
```
Comment (SarthakSavasil) [inline emotunes-commons/src/main/java/com/emotunes/emotunes/repository/UserSongEmotionPreferenceRepository.java]: list of tuple use krna chaiye wese, but agr work kr rha hai to chhod

File: emotunes-admin/src/main/java/com/emotunes/emotunes/service/impl/SchedulingServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        List<List<String>> userIdSongIdList = userSongEmotionPreferenceDao.getUserIdAndSongId(FETCH_COUNT_LIMIT);
+
+        List<String> songUrlListOfUser = new ArrayList<>();
+        for (int i = 0; i<userIdSongIdList.size(); i++) {
+            String userId = userIdSongIdList.get(i).get(0);
+            String songId = userIdSongIdList.get(i).get(1);
+            // todo: use caffeine cache to get url of songs
+            String songUrl = songsDao.getSongUrl(songId);
+
+            songUrlListOfUser.add(songUrl);
```
Comment (SarthakSavasil) [inline emotunes-admin/src/main/java/com/emotunes/emotunes/service/impl/SchedulingServiceImpl.java]: ye loop ke bahr hona chaiye, song ids ki list pass kr upar list of tuple nikal aur map mai convert kr as songid, urls

**Review by SarthakSavasil — CHANGES_REQUESTED** [review]

File: emotunes-admin/src/main/java/com/emotunes/emotunes/service/impl/SchedulingServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                        .distinct()
+                        .collect(Collectors.toList());
+
+        List<String> userIdList =
+                userIdSongIdList.stream()
+                        .map(tuple -> tuple.get(0).toString())
+                        .distinct()
+                        .collect(Collectors.toList());
+
+        List<String> songUrlList = songsDao.getSongUrls(songIdList);
```
Comment (SarthakSavasil) [inline emotunes-admin/src/main/java/com/emotunes/emotunes/service/impl/SchedulingServiceImpl.java]: song url direct fetch krke alag se map kiya hai iski jagah ek tuple fetch krte with (song_id, song_url) waha se map mai daalte.
Like abhi agr main query manipulate krta hoon ya pagination daalta hoon ya order by to ye logic galat map kr dega

Agr time lagega to ese hi rehne de

File: emotunes-admin/src/main/java/com/emotunes/emotunes/service/impl/SchedulingServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        Map<String, String> songIdSongUrlMap = mapStringLists(songIdList, songUrlList);
+
+        List<String> modelWeightsUrlList = userDao.getModelWeightsUrls(userIdList);
```
Comment (SarthakSavasil) [inline emotunes-admin/src/main/java/com/emotunes/emotunes/service/impl/SchedulingServiceImpl.java]: same

File: emotunes-commons/src/main/java/com/emotunes/emotunes/constants/AzureStorageConstans.java [hunk 7]
```diff
@@ -9,6 +9,9 @@ public class AzureStorageConstans {
 
     public static final String THUMBNAILS_CONTAINER = "thumbnails";
 
+
+    public static final String MODEL_WEIGHTS_CONTAINER = "thumbnails";
```
Comment (SarthakSavasil) [inline emotunes-commons/src/main/java/com/emotunes/emotunes/constants/AzureStorageConstans.java]: iski value thumbnail hi hogi?

File: emotunes-commons/src/main/java/com/emotunes/emotunes/repository/SongRepository.java [hunk 8]
```diff
@@ -21,4 +21,9 @@ public interface SongRepository extends JpaRepository<StoredSong, String> {
             nativeQuery = true
     )
     List<String> findPaginatedSongsByPrefix(String prefix, int pageSize, Long offset);
+
+    @Query(value =
+            "select song_url from song where id in ?1",
+            nativeQuery = true)
+    List<String> getSongUrls(List<String> songIdList);
```
Comment (SarthakSavasil) [inline emotunes-commons/src/main/java/com/emotunes/emotunes/repository/SongRepository.java]: is  jagah bol rha List<Tuple> tuple->songid, songUrl

**Review by SarthakSavasil — APPROVED** [review]

File: emotunes-commons/src/main/java/com/emotunes/emotunes/repository/SongRepository.java (hunk 8 above)
Comment (krishrustagi) [reply emotunes-commons/src/main/java/com/emotunes/emotunes/repository/SongRepository.java]: done

**Pull request merged**